oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/token/OAuth2AuthorizationGrantContext.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.server.authorization.token;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClient;
+import org.springframework.util.Assert;
+
+/**
+ * Context describing current status of OAuth2 authorization. User by a {@link OAuth2TokenIssuer} to issue a new OAuth2 token.
+ *
+ * Different implementations of {@link OAuth2TokenIssuer} may decide to use different properties of
+ * the context.
+ *
+ * @author Alexey Nesterov
+ * @since 0.1.0
+ */
+public class OAuth2AuthorizationGrantContext {
```

Comment:
I'm wondering if we should rename this to `OAuth2TokenContext`? It might be more intuitive given the other names: `OAuth2TokenIssuer`, `OAuth2TokenResult`, `OAuth2TokenMetadata` and `AbstractOAuth2Token` (`OAuth2Token`)

Reply:
What about renaming it back to `OAuth2TokenRequest`? I think of a context as something I can put data into and get data from. In this case this just a read-only object that a provider uses to pass data to token issuer. And the flow is quite logical then `OAuth2TokenRequest -> OAuth2TokenIssuer + OAuth2TokenCustomizer -> OAuth2TokenResult`.